pilot/pkg/security/authz/model/generator_test.go

```diff
@@ -455,6 +458,101 @@ func TestGenerator(t *testing.T) {
 	}
 }
 
+func TestServiceAccount(t *testing.T) {
+	input := "my-ns/my-sa"
+	cases := []struct {
+		Name     string
+		Identity string
+		Match    bool
+	}{
+		{
+			Name:     "standard",
+			Identity: "spiffe://cluster.local/ns/my-ns/sa/my-sa",
+			Match:    true,
+		},
+		{
+			Name:     "suffix attributes",
+			Identity: "spiffe://cluster.local/ns/my-ns/sa/my-sa/k/v",
+			Match:    true,
+		},
+		{
+			Name:     "prefix attributes",
+			Identity: "spiffe://cluster.local/k/v/ns/my-ns/sa/my-sa",
+			Match:    true,
+		},
+		{
+			Name:     "middle attributes",
+			Identity: "spiffe://cluster.local/ns/my-ns/k/v/sa/my-sa",
+			Match:    true,
+		},
+		{
+			Name:     "all attributes",
+			Identity: "spiffe://cluster.local/k1/v1/ns/my-ns/k2/v2/sa/my-sa/k3/v3",
+			Match:    true,
+		},
+		{
+			Name:     "sa suffix string",
+			Identity: "spiffe://cluster.local/ns/my-ns/sa/my-sa-suffix",
+			Match:    false,
+		},
+		{
+			Name:     "ns suffix string",
+			Identity: "spiffe://cluster.local/ns/my-ns-suffix/sa/my-sa",
+			Match:    false,
+		},
+		{
+			Name:     "not spiffe",
+			Identity: "cluster.local/ns/my-ns/sa/my-sa",
+			Match:    false,
+		},
+		{
+			Name:     "invalid spiffe",
+			Identity: "spiffe://ns/my-ns/sa/my-sa",
+			Match:    false,
+		},
+		{
+			Name:     "missing sa",
+			Identity: "spiffe://cluster.local/ns/my-ns",
+			Match:    false,
+		},
+		{
+			Name:     "missing ns",
+			Identity: "spiffe://cluster.local/sa/my-sa",
+			Match:    false,
+		},
+		{
+			Name:     "missing ns",
+			Identity: "spiffe://cluster.local/sa/my-sa",
+			Match:    false,
+		},
+		{
+			Name: "weird keys",
+			// This test case matches when it shouldn't ideally.
+			// Spiffe is a set of k/v pairs. Here we are accidentally matching
+			// on previous value + next key when we shouldn't.
+			// The chance of an identity being formatted like this is exceptionally low though
+			Identity: "spiffe://cluster.local/" + strings.Join([]string{
+				"k", "ns",
+				"my-ns", "something",
+				"bar", "sa",
+				"my-sa", "baz",
+			}, "/"),
+			Match: true,
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.Name, func(t *testing.T) {
+			r, _ := serviceAccountRegex(input)
+			// Parse as regex. Envoy does a full string match, so handle that
+			rgx, err := regexp.Compile("^" + r + "$")
+			if err != nil {
+				t.Fatal(err)
+			}
```

Comment:
I do think we need to explicitly use the SPIFFE validation library, at least in a test - the correctness of the regex needs to be checked against the actual SPIFFE spec library at some point.

If we don't want to do it at runtime (and it currently looks like nowhere in the `istio` codebase do we actually _check_ that the official SPIFFE library can parse or validate the SPIFFE IDs we're handling, which seems bad) , doing it in a test might be good.

Something like

```
if match == true, assert tt.Identity (parsable by SPIFFE lib)
```

Reply:
I initially agreed and went to do it and found we don't actually use this library today which gave me some pause to add a new dependency. Then I also realized this would be testing _our test cases_, not our code, which seems a bit redundant -- maybe worth it if it was free, but doesn't seem worth the extra dependency just for this. FWIW I did do a one-off run and as expected it validates all of our test cases other than the "not spiffe" case.